core/src/main/java/jenkins/websocket/WebSocketEcho.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2019 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.websocket;
+
+import hudson.Extension;
+import hudson.model.InvisibleAction;
+import hudson.model.UnprotectedRootAction;
+import java.nio.ByteBuffer;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+
+@Extension
+@Restricted(NoExternalUse.class)
+public class WebSocketEcho  extends InvisibleAction implements UnprotectedRootAction {
```

Comment:
I'm not understanding yet why this is needed.

Reply:
This is just here for testing and diagnostics—it is the most convenient way to check whether Jenkins is actually serving WebSocket connections, including through load balancers and whatnot, and to verify keepalive behavior. [JEP-222: _Interactive Tests_](https://github.com/jenkinsci/jep/blob/master/jep/222/README.adoc#interactive-tests)

Reply:
I'm a little concerned about potential security implications here. Not that I can enumerate any specific ones, though. It's just that this provides an unauthenticated channel. There probably isn't a a DOS attack, because there isn't any noticeable multiplier. Hmmm ... maybe it could be used to pump a lot of worthless, unauthenticated data through a Jenkins server. That sounds like a DOS attack.

Reply:
More than just hitting, say, `http://jenkins/login` repeatedly?

Reply:
I don't know. It seems like this allows someone to open a socket and just stream data through and back, which seems to be worse than just hitting the login page. Maybe it should just send back an abbreviated or limited length message. The first 10 characters or something.

Reply:
I cannot see that making any difference; one way or another it needs to handle the incoming message.

Maybe I will just restrict this to administrators. (Unlike TCP connections, it is trivial to pass authentication.)

Reply:
True. I do like the idea of restricting it to administrators. That removes concerns.

Reply:
[Serendipitously](https://github.com/vi/websocat/releases/tag/v1.5.0)

> Using client URI's like `websocat ws://user:password@host/` now adds basic authentication HTTP header to request

Reply:
That's nicely handy and timely. The behavior seems reasonable and works fine now.